lib/mlir/Dialect/Clift/Utils/ImportModel.cpp

```diff
@@ -0,0 +1,589 @@
+      return getTypeAttribute(T, RequireComplete);
+    }
+
+    case model::TypeKind::PrimitiveType:
+      revng_abort("Primitive types have no corresponding attribute.");
+
+    case model::TypeKind::Invalid:
+    case model::TypeKind::Count:
+      revng_abort("These are invalid values. Something has gone wrong.");
+    }
+  }
+
+  RecursiveCoroutine<clift::ValueType>
+  getUnwrappedValueType(const model::Type &ModelType,
+                        bool RequireComplete = false,
+                        const bool Const = false) {
+    if (const auto *const T = llvm::dyn_cast<model::PrimitiveType>(&ModelType))
+      rc_return getPrimitiveType(*T, Const);
+
+    const auto getDefinedType = [&](const auto Attr) -> clift::ValueType {
+      return make<clift::DefinedType>(Attr, getBool(Const));
+    };
+
+    if (const auto It = Cache.find(ModelType.ID()); It != Cache.end())
+      rc_return getDefinedType(It->second);
+
+    const clift::TypeDefinition Attr = getTypeAttribute(ModelType,
+                                                        RequireComplete);
+
+    if (not Attr)
+      rc_return nullptr;
+
+    if (RequireComplete) {
+      const auto R = Cache.try_emplace(ModelType.ID(), Attr);
+      revng_assert(R.second);
+    }
+
+    rc_return getDefinedType(Attr);
+  }
+
+  RecursiveCoroutine<clift::ValueType>
+  getQualifiedValueType(const model::QualifiedType &ModelType,
+                        bool RequireComplete = false) {
+    if (not ModelType.UnqualifiedType().isValid()) {
+      if (EmitError)
+        EmitError() << "Invalid UnqualifiedType in QualifiedType";
+      rc_return nullptr;
+    }
+
+    // Qualifier::isConst verifies itself, which interferes with unit testing.
+    const auto IsConst = [](const model::Qualifier &Q) {
+      return Q.Kind() == model::QualifierKind::Const;
+    };
+
+    // Qualifier::isPointer verifies itself, which interferes with unit testing.
+    const auto IsPointer = [](const model::Qualifier &Q) {
+      return Q.Kind() == model::QualifierKind::Pointer;
+    };
+
+    auto Qualifiers = llvm::ArrayRef(ModelType.Qualifiers());
+
+    const auto takeQualifier = [&]() -> const model::Qualifier & {
+      const model::Qualifier &Qualifier = Qualifiers.back();
+      Qualifiers = Qualifiers.slice(0, Qualifiers.size() - 1);
+      return Qualifier;
+    };
+
+    const auto takeConst = [&]() -> bool {
+      if (not Qualifiers.empty() and IsConst(Qualifiers.back())) {
+        Qualifiers = Qualifiers.slice(0, Qualifiers.size() - 1);
+        return true;
+      }
+      return false;
+    };
+
+    // If the set of qualifiers contains any pointers,
+    // the base type does not need to be complete.
+    if (RequireComplete and llvm::any_of(Qualifiers, IsPointer))
+      RequireComplete = false;
+
+    clift::ValueType ResultType = rc_recur
+      getUnwrappedValueType(*ModelType.UnqualifiedType().get(),
+                            RequireComplete,
+                            takeConst());
```

Comment:
Ah so, the "last" qualifier is pushed inside here and managed by `getUnwrappedValueType`.

Hopefully this will be simpler and less awkward with ivan's new patchset on model types.

Reply:
Yes, I reckon this will become much simpler.